_posts/2023-11-01-#28368.md

```diff
@@ -7,18 +7,79 @@ authors: [ismaelsadeeq]
+
+* Before this PR [`CTxMempool`](https://github.com/bitcoin/bitcoin/blob/22fa1f4702e9a60e141f545f8d09704deca34b22/src/txmempool.h#L301) owns a [`CBlockPolicyEstimator`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/policy/fees.h#L146C13-L146C13) as one of its members and, whenever a transaction is added or removed from the mempool in [`addUnchecked`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/txmempool.h#L475) and [`removedUnchecked`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/txmempool.h#L787) methods, it also updates the fee estimator synchronously.
+
+* Similarly, whenever a new block is connected, the [`removeForBlock`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/txmempool.h#L487) method removes the block transactions that are in the mempool together with its ancestors. This also triggers an update to the fee estimator synchronously.
+ 
+* Since `processBlock` happens within [`removeForBlock`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/txmempool.h#L487), these fee estimator updates blocks [`ConnectTip`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/validation.h#L732C13-L732C13). Adding more steps to the fee estimator during this update can slow down block processing even further.
+
+* Subclasses of [`CValidationInterface`](https://github.com/bitcoin/bitcoin/blob/9c30f5ef9d6712539cb8dc0f52b3a2206a25e074/src/validationinterface.h#L77) can implement its methods, and specify  the instructions to be executed in response to the corresponding event. 
+and register the subclass with [`CMainSignals`](https://github.com/bitcoin/bitcoin/blob/9c30f5ef9d6712539cb8dc0f52b3a2206a25e074/src/validationinterface.h#L182C7-L182C19) using [`RegisterValidationInterface`](https://github.com/bitcoin/bitcoin/blob/9c30f5ef9d6712539cb8dc0f52b3a2206a25e074/src/validationinterface.h#L26).
+Validation and mempool use [`CMainSignals`](https://github.com/bitcoin/bitcoin/blob/9c30f5ef9d6712539cb8dc0f52b3a2206a25e074/src/validationinterface.h#L182C7-L182C19) to "notify" subscribers of these events. 
+The "notification" may entail executing a callback immediately or adding it to a queue of callbacks to be executed in the background. 
+* These Subclasses can be called clients or subscribers of the [`CValidationInterface`](https://github.com/bitcoin/bitcoin/blob/9c30f5ef9d6712539cb8dc0f52b3a2206a25e074/src/validationinterface.h#L77).
+
+    * These events are executed asynchronously on a different thread from the one that fired the event.
```

Comment:
Do you mean callback or event? Also, just a few lines up you mention that "The "notification" may entail executing a callback immediately or adding it to a queue of callbacks to be executed in the background. ", that seems contradictory?